efi/preinstall/check_tcglog.go

```diff
@@ -0,0 +1,610 @@
+	return t.phase == tcglogPhaseOSPresent
+}
+
+// checkFirmwareLogAndChoosePCRBank verifies that the firmware TCG log is in crypto-agile form and
+// consistent with at least one supported PCR bank for the specified mandatory PCRs when reconstructed
+// for the TCG defined PCRs (0-7). It also ensures that:
+//   - the TCG defined PCRs contain a EV_SEPARATOR event between the pre-OS and OS-present environment (
+//     although the one in PCR7 separates secure boot configuratuib from secure boot authentication).
+//   - that none of the EV_SEPARATORs in the TCG defined PCRs indicated that an error occurred.
+//   - there are no pre-OS measurements to non-TCG defined PCRs (8-).
+//
+// This won't return an error for failures in TCG defined PCRs if they aren't part of the specified mandatory
+// PCRs set, but the errors will be accessible on the returned results struct.
+//
+// The returned results struct indicates the best PCR bank to use and specifies the TPM startup locality as well.
+func checkFirmwareLogAndChoosePCRBank(tpm *tpm2.TPMContext, log *tcglog.Log, mandatoryPcrs tpm2.HandleList) (results *pcrBankResults, err error) {
+	// Make sure it's a crypto-agile log
+	if !log.Spec.IsEFI_2() {
+		return nil, errors.New("invalid log spec")
+	}
+
+	// Chose the best PCR bank, ordered from SHA-512, SHA-384 to SHA-256. We're most
+	// likely to get SHA-256 here - it's only in very recent devices that we have TPMs with
+	// SHA-384 support and corresponding firmware integration.
+	// We try to keep all errors enountered during selection here.
+	mainErr := new(NoSuitablePCRAlgorithmError)
+	var chosenResults *pcrBankResults
+	for _, alg := range supportedAlgs {
+		if chosenResults != nil {
+			// We've already got a good PCR bank, so no need to carry on.
+			break
+		}
+
+		results, err := checkFirmwareLogAgainstTPMForAlg(tpm, log, alg, mandatoryPcrs)
+		switch {
+		case err != nil:
+			// This entire bank is bad
+			mainErr.setBankErr(alg, err)
+		case results.Ok():
+			// This is a good PCR bank
+			chosenResults = results
+		default:
+			// This isn't a good PCR bank because some mandatory PCRs
+			// failed. Record the individual PCR errors.
+			mainErr.setPcrErrs(results)
+		}
+	}
+
+	if chosenResults == nil {
+		// No suitable PCR bank was found, so return an error that's hopefully useful :(
+		return nil, mainErr
+	}
+
+	// Make sure that the log is well formed and has well defined phases and that there are no measurements
+	// to non-TCG defined PCRs until OS-present.
+	phaseTracker := newTcgLogPhaseTracker()
+	for _, ev := range log.Events {
+		phase, err := phaseTracker.processEvent(ev)
+		if err != nil {
+			return nil, err
+		}
+		if phase == tcglogPhaseOSPresent {
+			// We've hit OS-present - we don't care about the rest of the log at
+			// this stage
+			break
+		}
+	}
+	if !phaseTracker.reachedOSPresent() {
+		return nil, errors.New("reached the end of the log without seeing EV_SEPARATOR events in all TCG defined PCRs")
```

Comment:
this error case is not tested

Reply:
This one should be easy to figure out how to test.

Reply:
I've added a test for this one now.